File: javascript/leetcode75/array-strings/49. Group Anagrams.js
class Solution {
    groupAnagrams(strs) {
        let res = {};
        for (let word of strs) {
            let temp = word.split('').sort().join('');
            if (temp in res) {
                res[temp].push(word);
            } else {
                res[temp] = [word];
            }
        }
        let result = [];
        for (let x in res) {
            result.push(res[x]);
        }
        return result;
    }
}

let strs = ["eat", "tea", "tan", "ate", "nat", "bat"];
let s = new Solution();
console.log(s.groupAnagrams(strs));

// Input: strs = ["eat","tea","tan","ate","nat","bat"]
// Output: [["bat"],["nat","tan"],["ate","eat","tea"]] 

File: python/patterns/hash-map/49. Group Anagrams.py
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = {}
        for word in strs:
            temp = ''.join(sorted(word))
            if temp in res:
                res[temp].append(word)
            else:
                res[temp] = [word]
        result =[]
        for x in res:
            result.append(res[x])
        return result


        
strs = ["eat","tea","tan","ate","nat","bat"]
s = Solution()
print(s.groupAnagrams(strs))

"""
Time complexity: O(N * K log K)
Space complexity: O(N * K)
where N is the number of strings and K is the maximum length of any string.
"""



# Input: strs = ["eat","tea","tan","ate","nat","bat"]
# Output: [["bat"],["nat","tan"],["ate","eat","tea"]]